### PR TITLE
Implementation of InstanceIdClient to allow subscribe/unsubscribe to topic

### PR DESCRIPTION
**Pull request by Leo-Mepham**
As per #93 (referencing #83 )
Contributor License Agreement signed
Unit Tests & Integration Tests passing.

This is a conversion of the Java file at https://github.com/firebase/firebase-admin-java/blob/master/src/main/java/com/google/firebase/messaging/InstanceIdClientImpl.java and incorporating that into the main FirebaseMessaging class.

Look forward to the review :)

### CODE REVIEW

**Comment by Leo-Mepham** [comment]
Have done another CLA - Forgot my git config email address is different to my github email address.

File: FirebaseAdmin/FirebaseAdmin/Messaging/InstanceIdClient.cs [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+    /// <param name="topic">The topic name to unsubscribe from. /topics/ will be prepended to the topic name provided if absent.</param>
+    /// <param name="registrationTokens">A list of registration tokens to unsubscribe.</param>
+    /// <returns>A task that completes with a <see cref="TopicManagementResponse"/>, giving details about the topic unsubscription operations.</returns>
+    public async Task<TopicManagementResponse> UnsubscribeFromTopicAsync(string topic, List<string> registrationTokens)
+    {
+        try
+        {
+            return await this.SendInstanceIdRequest(topic, registrationTokens, IidUnsubscribePath).ConfigureAwait(false);
+        }
+        catch (HttpRequestException e)
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin/Messaging/InstanceIdClient.cs]: SendInstancedIdRequest already does the error handling. This shouldn't be needed.

File: FirebaseAdmin/FirebaseAdmin/Messaging/InstanceIdClient.cs [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="InstanceIdClient"/> class.
+    /// </summary>
+    /// <param name="clientFactory">A default implentation of the HTTP client factory.</param>
+    /// <param name="credential">An instance of the <see cref="GoogleCredential"/> GoogleCredential class.</param>
+    /// <param name="projectId">The Project Id for FCM Messaging.</param>
+    public InstanceIdClient(HttpClientFactory clientFactory, GoogleCredential credential, string projectId)
+    {
+        if (string.IsNullOrEmpty(projectId))
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin/Messaging/InstanceIdClient.cs]: projectId is not needed here.

File: FirebaseAdmin/FirebaseAdmin/Messaging/InstanceIdClient.cs [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+
+    private class InstanceIdServiceRequest
+    {
+        [JsonProperty("to")]
+        public string Topic { get; set; }
+
+        [JsonProperty("registration_tokens")]
+        public List<string> RegistrationTokens { get; set; }
+    }
+}
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin/Messaging/InstanceIdClient.cs]: Add new line

File: FirebaseAdmin/FirebaseAdmin/Messaging/InstanceIdClient.cs [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+        };
+
+        request.Headers.Add("access_token_auth", "true");
+
+        try
+        {
+            var response = await this.httpClient.SendAsync(request, default(CancellationToken)).ConfigureAwait(false);
+            var json = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
+            this.errorHandler.ThrowIfError(response, json);
+            return JsonConvert.DeserializeObject<TopicManagementResponse>(json);
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin/Messaging/InstanceIdClient.cs]: This doesn't look right. You probably need a new `InstanceIfServiceResponse` type, to which the response can be deserialized. Then you need to create a `TopicManagementResponse` off of that.

File: FirebaseAdmin/FirebaseAdmin.Tests/Messaging/TopicManagementResponseTest.cs [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+using System.Collections.Generic;
+using Newtonsoft.Json.Linq;
+using Xunit;
+
+namespace FirebaseAdmin.Tests.Messaging
+{
+    public class TopicManagementResponseTest
+    {
+        [Fact]
+        public void SuccessfulTopicManagementResponse()
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin.Tests/Messaging/TopicManagementResponseTest.cs]: `TopicManagmentResponse` is redundant in these test case names. I recommend just shortening to "SuccessfulResponse", "UnsuccessfulResponse", "NullResponse" and so on.

File: FirebaseAdmin/FirebaseAdmin.Tests/Messaging/TopicManagementResponseTest.cs [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+            Assert.Equal(1, response.SuccessCount);
+        }
+
+        [Fact]
+        public void TopicManagementResponseCannotBeNull()
+        {
+            Assert.Throws<ArgumentException>(() =>
+            {
+                List<JObject> jObjects = null;
+                var response = new TopicManagementResponse(jObjects);
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin.Tests/Messaging/TopicManagementResponseTest.cs]: Just `new TopicManagementResponse(null);`

File: FirebaseAdmin/FirebaseAdmin.Tests/Messaging/TopicManagementResponseTest.cs [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    public class TopicManagementResponseTest
+    {
+        [Fact]
+        public void SuccessfulTopicManagementResponse()
+        {
+            var json = @"[{}, {}]";
+            var jObjects = JArray.Parse(json).ToObject<List<JObject>>();
+            var response = new TopicManagementResponse(jObjects);
+
+            Assert.Equal(0, response.FailureCount);
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin.Tests/Messaging/TopicManagementResponseTest.cs]: Should also run some assertions on `response.Errors`. Here and in other tests.

File: FirebaseAdmin/FirebaseAdmin.Tests/Messaging/InstanceIdClientTest.cs [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+        {
+            var clientFactory = new HttpClientFactory();
+            Assert.Throws<ArgumentException>(
+                () => new InstanceIdClient(clientFactory, MockCredential, null));
+            Assert.Throws<ArgumentException>(
+                () => new InstanceIdClient(clientFactory, MockCredential, string.Empty));
+        }
+
+        [Fact]
+        public void InstanceIdClientThrowsOnNoCredential()
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin.Tests/Messaging/InstanceIdClientTest.cs]: Shorten the names by dropping redundant labels -- "NoCredential", "NoClientFactory", "SubscribeToTopic" etc

File: FirebaseAdmin/FirebaseAdmin.Tests/Messaging/FirebaseMessagingTest.cs [hunk 9]
```diff
@@ -118,6 +119,56 @@ public async Task SendMessageCancelWithClientFactory()
                     new Message() { Topic = "test-topic" }, canceller.Token));
         }
 
+        [Fact]
+        public async Task SubscribeWithClientFactory()
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin.Tests/Messaging/FirebaseMessagingTest.cs]: SubscribeToTopicAsync

File: FirebaseAdmin/FirebaseAdmin.Tests/Messaging/FirebaseMessagingTest.cs [hunk 10]
```diff
@@ -118,6 +119,56 @@ public async Task SendMessageCancelWithClientFactory()
...
+            Assert.Same(messaging, FirebaseMessaging.GetMessaging(app));
+
+            var response = await messaging.SubscribeToTopicAsync("test-topic", new List<string> { "test-token" });
+            Assert.Equal(0, response.FailureCount);
+            Assert.Equal(1, response.SuccessCount);
+            app.Delete();
+        }
+
+        [Fact]
+        public async Task UnsubscribeWithClientFactory()
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin.Tests/Messaging/FirebaseMessagingTest.cs]: UnsubscribeFromTopicAsync

**Comment by Leo-Mepham** [comment]
Thanks for the review @hiranya911 a lot of good suggestions, I'm grateful for that and have implemented those. Let me know what you think about the new changes?

I also wanted to ask - I've not yet built and run this against a real application (IE: Subscribing my personal phone to a topic using this code, then sending a message to the topic and receiving it) - Is that something I should do, or maintainers do, or is there another practice commonly followed?

File: FirebaseAdmin/FirebaseAdmin.Tests/Messaging/TopicManagementResponseTest.cs [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+        [Fact]
+        public void UnsuccessfulResponse()
+        {
+            var json = @"{""results"": [{}, {""error"":""NOT_FOUND""}]}";
+            var instanceIdServiceResponse = JsonConvert.DeserializeObject<InstanceIdServiceResponse>(json);
+            var response = new TopicManagementResponse(instanceIdServiceResponse);
+
+            Assert.Equal(1, response.FailureCount);
+            Assert.Equal(1, response.SuccessCount);
+            Assert.NotEmpty(response.Errors);
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin.Tests/Messaging/TopicManagementResponseTest.cs]: Can we do Assert.Single here?

File: FirebaseAdmin/FirebaseAdmin.Tests/Messaging/TopicManagementResponseTest.cs [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+            Assert.Equal("registration-token-not-registered", response.Errors[0].Reason);
+            Assert.Equal(1, response.Errors[0].Index);
+        }
+
+        [Fact]
+        public void NullResponse()
+        {
+            Assert.Throws<ArgumentException>(() =>
+            {
+                var response = new TopicManagementResponse(null);
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin.Tests/Messaging/TopicManagementResponseTest.cs]: Nit: Remove `var response =` part.

File: FirebaseAdmin/FirebaseAdmin.Tests/Messaging/TopicManagementResponseTest.cs [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+            });
+        }
+
+        [Fact]
+        public void EmptyResponse()
+        {
+            Assert.Throws<ArgumentException>(() =>
+            {
+                var instanceIdServiceResponse = new InstanceIdServiceResponse();
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin.Tests/Messaging/TopicManagementResponseTest.cs]: Move this line outside the `Throws` block. Remove `var response =` part.

File: FirebaseAdmin/FirebaseAdmin/Messaging/InstanceIdClient.cs [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+
+        /// <summary>
+        /// Dispose the HttpClient.
+        /// </summary>
+        public void Dispose()
+        {
+            this.httpClient.Dispose();
+        }
+
+        private async Task<TopicManagementResponse> SendInstanceIdRequest(string topic, List<string> registrationTokens, string path)
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin/Messaging/InstanceIdClient.cs]: We need to validate `registrationTokens`:

1. Must not be null or empty
2. Length must not exceed 1000
3. Each entry in the list must not be null or empty

File: FirebaseAdmin/FirebaseAdmin/Messaging/InstanceIdServiceResponse.cs [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+using System.Linq;
+using Newtonsoft.Json;
+
+namespace FirebaseAdmin.Messaging
+{
+    /// <summary>
+    /// Response from an operation that subscribes or unsubscribes registration tokens to a topic.
+    /// See <see cref="FirebaseMessaging.SubscribeToTopicAsync(string, List{string})"/> and <see cref="FirebaseMessaging.UnsubscribeFromTopicAsync(string, List{string})"/>.
+    /// </summary>
+    public sealed class InstanceIdServiceResponse
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin/Messaging/InstanceIdServiceResponse.cs]: Lets make this another private nested class of `InstanceIdClient`. Definitely shouldn't be public.

File: FirebaseAdmin/FirebaseAdmin/Messaging/TopicManagementResponse.cs [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+        /// <summary>
+        /// Gets a list of errors encountered while executing the topic management operation.
+        /// </summary>
+        /// <returns>A non-null list.</returns>
+        public IReadOnlyList<Error> Errors { get; private set; }
+
+        /// <summary>
+        /// A topic management error.
+        /// </summary>
+        public sealed class Error
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin/Messaging/TopicManagementResponse.cs]: Our API review team has agreed that should be called `ErrorInfo` and it should be a top-level (i.e. un-nested) class. Lets rename and move to a new file.

File: FirebaseAdmin/FirebaseAdmin.Tests/Messaging/InstanceIdClientTest.cs [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+            };
+            var factory = new MockHttpClientFactory(handler);
+
+            var client = new InstanceIdClient(factory, MockCredential);
+
+            var result = await client.UnsubscribeFromTopicAsync("test-topic", new List<string> { "abc123" });
+
+            Assert.Equal(1, result.SuccessCount);
+        }
+    }
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin.Tests/Messaging/InstanceIdClientTest.cs]: Can we add a couple more test cases here? Specifically to test what happens when the server responds with a non-200 response.

**Comment by Leo-Mepham** [comment]
Hi @hiranya911 , I've made the changes you suggested, thanks again for your feedback.

Making the InstanceIdServiceResponse class a private class of InstanceIdClient means I can't provide it to the constructor of TopicManagementResponse (Unless I've missed a trick here?) so I have passed a list of all the messages from the InstanceIdServiceResponse instead. Let me know if you think there is a better approach. I had wanted to convert to a List<ErrorInfo> inside the InstanceIdClient and pass that to TopicManagementResponse constructor the but then I can't provide the index of the error back to the library caller which means they can't see which token caused the error.

I've added in HTTP status code error tests as requested, they all throw FirebaseMessagingException which looks to come from the FirebaseAdmin.Messaging.HttpErrorHandler so should give something useful to the caller.

Looking forward to the feedback again, let me know what you think :)

File: FirebaseAdmin/FirebaseAdmin.Tests/Messaging/InstanceIdClientTest.cs [hunk 18]
```diff
@@ -60,5 +61,85 @@ public async Task UnsubscribeFromTopicAsync()
...
+            {
+                StatusCode = HttpStatusCode.Unauthorized,
+                Response = "Unauthorized",
+            };
+            var factory = new MockHttpClientFactory(handler);
+
+            var client = new InstanceIdClient(factory, MockCredential);
+
+            await Assert.ThrowsAsync<FirebaseMessagingException>(
+               () => client.SubscribeToTopicAsync("test-topic", new List<string> { "abc123" }));
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin.Tests/Messaging/InstanceIdClientTest.cs]: You should be able to run the following assertions on the exception:

```
Assert.Equal(ErrorCode.Unauthenticated, exception.ErrorCode);
Assert.Equal("Unexpected HTTP response with status: 401 (Unauthorized)\nUnauthorized", exception.Message);
Assert.Null(exception.MessagingErrorCode);
Assert.NotNull(exception.HttpResponse);
Assert.Null(exception.InnerException);
```

Similar assertions should be possible in the following test cases as well.

File: FirebaseAdmin/FirebaseAdmin/Messaging/ErrorInfo.cs [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+namespace FirebaseAdmin.Messaging
+{
+    /// <summary>
+    /// A topic management error.
+    /// </summary>
+    public sealed class ErrorInfo
+    {
+        // Server error codes as defined in https://developers.google.com/instance-id/reference/server
+        // TODO: Should we handle other error codes here (e.g. PERMISSION_DENIED)?
+        private static IReadOnlyDictionary<string, string> errorCodes;
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin/Messaging/ErrorInfo.cs]: Let's assign this dictionary here (i.e. move that logic out of the constructor). Also make it `readonly`.

File: FirebaseAdmin/FirebaseAdmin/Messaging/ErrorInfo.cs [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+        private readonly string unknownError = "unknown-error";
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin/Messaging/ErrorInfo.cs]: static

File: FirebaseAdmin/FirebaseAdmin/Messaging/InstanceIdClient.cs [hunk 21]
```diff
@@ -131,6 +141,32 @@ private FirebaseMessagingException CreateExceptionFromResponse(HttpRequestExcept
                 response: temp.HttpResponse);
         }
 
+        private void ValidateRegistrationTokenList(List<string> registrationTokens)
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin/Messaging/InstanceIdClient.cs]: This function should throw `ArgumentException` and `ArgumentNullException`.

File: FirebaseAdmin/FirebaseAdmin/Messaging/InstanceIdClient.cs [hunk 22]
```diff
@@ -151,5 +187,23 @@ private class InstanceIdServiceRequest
             [JsonProperty("registration_tokens")]
             public List<string> RegistrationTokens { get; set; }
         }
+
+        private class InstanceIdServiceResponse
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin/Messaging/InstanceIdClient.cs]: Ah... I see what you mean. You're also using this in `TopicManagementResponse`. In that case lets move this to its own file (like you had it earlier), but lets make it `internal`.

File: FirebaseAdmin/FirebaseAdmin/Messaging/TopicManagementResponse.cs [hunk 23]
```diff
@@ -25,21 +26,26 @@ public sealed class TopicManagementResponse
         /// <summary>
         /// Initializes a new instance of the <see cref="TopicManagementResponse"/> class.
         /// </summary>
-        /// <param name="instanceIdServiceResponse">The results from the response produced by FCM topic management operations.</param>
-        public TopicManagementResponse(InstanceIdServiceResponse instanceIdServiceResponse)
+        /// <param name="topicManagementResults">The results from the response produced by FCM topic management operations.</param>
+        public TopicManagementResponse(List<string> topicManagementResults)
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin/Messaging/TopicManagementResponse.cs]: Ok, I see why `InstanceIdServiceResponse` needs to be visible. Lets make that `internal` and change this constructor back to accept an `InstanceIdServiceResponse`.

File: FirebaseAdmin/FirebaseAdmin/Messaging/InstanceIdClient.cs [hunk 24]
```diff
@@ -145,24 +144,24 @@ private void ValidateRegistrationTokenList(List<string> registrationTokens)
...
-                throw new FirebaseMessagingException(ErrorCode.InvalidArgument, "Registration token list must not contain more than 1000 tokens");
+                throw new ArgumentException("Registration token list must not contain more than 1000 tokens");
             }
 
             foreach (var registrationToken in registrationTokens)
             {
                 if (string.IsNullOrEmpty(registrationToken))
                 {
-                    throw new FirebaseMessagingException(ErrorCode.InvalidArgument, "Registration token must not be null");
+                    throw new ArgumentNullException("Registration tokens must not be null");
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin/Messaging/InstanceIdClient.cs]: Throw `ArgumentException` and change the text to "Registration tokens must not be null or empty"

File: FirebaseAdmin/FirebaseAdmin/Messaging/InstanceIdClient.cs [hunk 25]
```diff
@@ -118,8 +118,7 @@ private async Task<TopicManagementResponse> SendInstanceIdRequest(string topic,
                     throw new ArgumentException("unexpected response from topic management service");
                 }
 
-                var results = instanceIdServiceResponse.Results.Select(r => r.Error).ToList();
-                return new TopicManagementResponse(results);
+                return new TopicManagementResponse(instanceIdServiceResponse);
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin/Messaging/InstanceIdClient.cs]: It's probably cleaner to move the above if condition to the constructor of `TopicManagementResponse`.

PS: It seems you already have it there. Remove the redundant check from here.

File: FirebaseAdmin/FirebaseAdmin/Messaging/InstanceIdServiceResponse.cs [hunk 26]
```diff
@@ -0,0 +1,46 @@
...
+
+        /// <summary>
+        /// Gets the number of results returned by the operation.
+        /// </summary>
+        public int ResultCount => Results?.Count() ?? 0;
+
+        /// <summary>
+        /// An instance Id response error.
+        /// </summary>
+        public class InstanceIdServiceResponseElement
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin/Messaging/InstanceIdServiceResponse.cs]: internal

**Comment by Leo-Mepham** [comment]
Thanks again for all you time, help, and patience here @hiranya911 

I have a couple of questions if that's OK?

What sort of time does it usually take for this to end up in the nuget package?

Once this goes in, can I call myself a "Google code contributor" or "Google contributor" or is there another term for people who add to this project :)

**Comment by hiranya911** [comment]
Thanks again for the pull request @Leo-Mepham. I'll try to get this included in this week's release. If not definitely next week.

I don't think we have an official term for the developers who contribute to our open source projects :) @samtstern do you have any thoughts on that?

In any case, please note that we are really grateful for your contribution, and we will make a note of that in the release notes when this goes out. I look forward to receiving more contributions from you in the future :)

**Comment by hiranya911** [comment]
@Leo-Mepham I'm told it's ok to refer to developers like you as "Google contributors". But please be sure to not represent yourself as a Google employee :)

**Comment by Leo-Mepham** [comment]
Great stuff, thanks @hiranya911 👍

**Pull request merged**